model/infinispan/src/main/java/org/keycloak/spi/infinispan/impl/embedded/JGroupsConfigurator.java

```diff
@@ -201,12 +211,48 @@ private static void configureDiscovery(ConfigurationBuilderHolder holder, Keyclo
         var tableName = JpaUtils.getTableNameForNativeQuery("JGROUPS_PING", em);
         var stack = getProtocolConfigurations(tableName, isUdp, tracingEnabled);
         var connectionFactory = (JpaConnectionProviderFactory) session.getKeycloakSessionFactory().getProviderFactory(JpaConnectionProvider.class);
-        holder.addJGroupsStack(new JpaFactoryAwareJGroupsChannelConfigurator(stackName, stack, connectionFactory, isUdp), null);
+
+        Address address = Retry.call(ignored -> KeycloakModelUtils.runJobInTransactionWithResult(session.getKeycloakSessionFactory(), JGroupsConfigurator::prepareJGroupsAddress), 50, 10);
+        holder.addJGroupsStack(new JpaFactoryAwareJGroupsChannelConfigurator(stackName, stack, connectionFactory, isUdp, address), null);
 
         transportOf(holder).stack(stackName);
         JGroupsConfigurator.logger.info("JGroups JDBC_PING discovery enabled.");
     }
 
+    /**
+     * Generate the next sequence of the address, and place it into the JGROUPS_PING table so other nodes can see it.
+     * If we are the first = smallest entry, the other nodes will wait for us to become a coordinator
+     * for max_join_attempts x all_clients_retry_timeout = 10 x 100 ms = 1 second. Otherwise, we will wait for that
+     * one second. This prevents a split-brain scenario on a concurrent startup.
+     */
+    private static Address prepareJGroupsAddress(KeycloakSession session) {
+        var storage = session.getProvider(ServerConfigStorageProvider.class);
+        String seq = storage.loadOrCreate(JGROUPS_ADDRESS_SEQUENCE, () -> "0");
+        long value = Long.parseLong(seq) + 1;
+        String newSeq = Long.toString(value);
+        storage.replace(JGROUPS_ADDRESS_SEQUENCE, seq, newSeq);
```

Comment:
Don't we need to handle the case where the replace fails, otherwise won't we potentially generate two addresses on different nodes with the same seq number?